src/wrapper/wrapper.c

```diff
@@ -216,15 +158,15 @@ real_call:;
     return (func_ptr (path, oflag, mode));
 }
 
-FILE *fopen (const char *path, const char *mode)
+DYAD_DLL_EXPORTED FILE *fopen (const char *path, const char *mode)
 {
     char *error = NULL;
     typedef FILE *(*fopen_ptr_t) (const char *, const char *);
     fopen_ptr_t func_ptr = NULL;
 
-    if (ctx == NULL) {
-        dyad_wrapper_init ();
-    }
+    // if (ctx == NULL) {
+    //     dyad_wrapper_init ();
+    // }
 
     func_ptr = (fopen_ptr_t)dlsym (RTLD_NEXT, "fopen");
```

Comment:
All this logic of moving to next symbol can be abstracted using macro for each case. 
```
#define MAP_OR_FAIL(func_)                                         \
  if (!(real_##func_##_)) {                                        \
    real_##func_##_ = (real_t_##func_##_)dlsym(RTLD_NEXT, #func_); \
    if (!(real_##func_##_)) {                                      \
      fprintf(stderr, "failed to map symbol\n");    \
    }                                                              \
  }
#else
#define MAP_OR_FAIL(func)
#endif
```

I recommend using GOTCHA tool which handles this more dynamically.

Reply:
I agree that switching the GOTCHA in the future is the right move. That definitely doesn't belong in this PR though. If we don't have one yet, I'll create an issue for switching from "raw" `LD_PRELOAD` to GOTCHA.

Reply:
Issue #43 has been created to track adding GOTCHA support

Reply:
It make sense to address it in a separate PR. @hariharan-devarajan please find time with us to explain the benefits from the users perspective as well as developer's.